include/scheduler.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef GRAPH_PROTOTYPE_SCHEDULER_HPP
+#define GRAPH_PROTOTYPE_SCHEDULER_HPP
+#include <graph.hpp>
+#include <set>
+#include <queue>
+
+namespace fair::graph::scheduler {
+
+struct init_proof {
+    init_proof(bool _success) : success(_success) {}
+    init_proof(init_proof && init): init_proof(init.success) {}
+    bool success = true;
+
+    init_proof& operator=(init_proof &&init)  noexcept {this->success = init; return *this;}
+    operator bool() const { return success; }
+};
+
+init_proof init(fair::graph::graph &graph) {
+    auto result = init_proof(
+            std::all_of(graph._connection_definitions.begin(), graph._connection_definitions.end(), [] (auto& connection_definition) {
+                return connection_definition() == connection_result_t::SUCCESS;
+            }));
+    graph._connection_definitions.clear();
+    return result;
+}
+
+/**
+ * Trivial loop based scheduler, which iterates over all nodes in definition order in the graph until no node did any processing
+ */
+class simple : public node<simple>{
+public:
+    explicit simple(fair::graph::graph &graph)  : init{fair::graph::scheduler::init(graph)}, graph(graph) { }
+
+    work_return_t work() {
+        if (!init) {
+            return work_return_t::ERROR;
+        }
+        bool run = true;
+        while (run) {
+            bool something_happened = false;
+            for (auto &node : graph._nodes) {
+                auto result = node->work();
+                if (result == work_return_t::ERROR) {
+                    return work_return_t::ERROR;
+                } else if (result == work_return_t::INSUFFICIENT_INPUT_ITEMS) {
+                    // nothing
+                } else if (result == work_return_t::DONE) {
+                    // nothing
+                } else if (result == work_return_t::OK) {
+                    something_happened = true;
+                } else if (result == work_return_t::INSUFFICIENT_OUTPUT_ITEMS) {
+                    something_happened = true;
+                }
+            }
+            run = something_happened;
+        }
+
+        return work_return_t::DONE;
+    }
+private:
```

Comment:
member fields on top, followed by public constructor ... IMO you may skip the `private:` declarator since fields after the class header definition are always private.